sled-agent/src/bootstrap/agent.rs

```diff
@@ -236,91 +408,154 @@ impl Agent {
-            // doesn't match ours. TODO-correctness Do we need to handle a
-            // partially-initialized rack where we may have a share from a
-            // previously-started-but-not-completed init process? If rerunning
-            // it produces different shares this check will fail.
-            if *trust_quorum_share != *self.share.lock().await {
-                let err_str = concat!(
-                    "Sled Agent already running with",
-                    " a different trust quorum share"
-                )
-                .to_string();
-                return Err(BootstrapError::SledError(err_str));
-            }
+                // Stop the bootstrap agent from monitoring for hardware, and
+                // pass control of service management to the sled agent.
+                //
+                // NOTE: If we fail at any point in the body of this function,
+                // we should restart the hardware monitor, so we can react to
+                // changes in the switch regardless of the success or failure of
+                // this sled agent.
+                let (hardware, services) = hardware_monitor
+                    .take()
+                    .expect("Hardware Monitor does not exist")
+                    .stop()
+                    .await
+                    .expect("Failed to stop hardware monitor");
+
+                // This acts like a "run-on-drop" closure, to restart the
+                // hardware monitor in the bootstrap agent if we fail to
+                // initialize the Sled Agent.
+                //
+                // In the "healthy" case, we can "cancel" and this is
+                // effectively a no-op.
+                struct RestartMonitor<'a> {
+                    run: bool,
+                    log: Logger,
+                    hardware: Option<HardwareManager>,
+                    services: Option<ServiceManager>,
+                    monitor: &'a mut Option<HardwareMonitor>,
+                }
+                impl<'a> RestartMonitor<'a> {
+                    fn cancel(mut self) {
+                        self.run = false;
+                    }
+                }
+                impl<'a> Drop for RestartMonitor<'a> {
+                    fn drop(&mut self) {
+                        if self.run {
+                            *self.monitor = Some(HardwareMonitor::start(
+                                &self.log,
+                                self.hardware.take().unwrap(),
+                                self.services.take().unwrap(),
+                            ));
+                        }
+                    }
+                }
+                let restarter = RestartMonitor {
+                    run: true,
+                    log: self.log.clone(),
+                    hardware: Some(hardware),
+                    services: Some(services.clone()),
+                    monitor: hardware_monitor,
+                };
```

Comment:
I fully admit, this is a little weird, but it felt like a safe way of solving this problem.

Basically: we need to stop the hardware monitor and take resources away from the bootstrap server to hand them off to the sled agent. But if this fails, we need to put them back.

I did consider changing the initialization of the sled agent, to make the `ServiceManager` an `Option`, but then sled agent would need error-handling for an "optional" `ServiceManager` while serving requests on its HTTP service, which seems grosser (and more visibly bad).

Reply:
I think I understand what's going on here. To check:

- `restarter` is capturing the hardware monitor (and some other stuff)
- Importantly, the `run` field is `true` at construction.
- If this method completes successfully, we `cancel()` the `restarter`, which sets `run` to false, and continue with...other stuff.
- If the method fails, `restarter` is dropped. Because `run` is true at construction, this will ensure the hardware monitor continues in that case.

So far so good. I think there's one thing I'm missing: don't we need to put the monitor "back" somewhere? If this method fails, how does the monitor not get dropped anyway?

Reply:
Ok, I think I see now. It's a separate Tokio task, which we signal to exit when we're satisfied.